File: PerDayOneQuestion/430_扁平化多级双向链表.py
"""
-------------------------------------
# -*- coding: utf-8 -*-
# @Time    : 2021/9/24 11:05:30
# @Author  : Giyn
# @Email   : giyn.jy@gmail.com
# @File    : 430_扁平化多级双向链表.py
# @Software: PyCharm
-------------------------------------
"""


class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child


class Solution:
    def flatten(self, head):
        def dfs(node):
            if not node:
                return
            if not node.child and not node.next:
                return node
            elif node.child:
                last = dfs(node.child)
                if last:
                    last.next = node.next
                if node.next:
                    node.next.prev = last
                node.next = node.child
                node.child.prev = node
                node.child = None
                return dfs(last)
            else:
                return dfs(node.next)

        dfs(head)

        return head


if __name__ == '__main__':
    solution = Solution()

File: linked-list/430_扁平化多级双向链表/430_扁平化多级双向链表.go
/*
-------------------------------------
# @Time    : 2022/3/11 19:24:40
# @Author  : Giyn
# @Email   : giyn.jy@gmail.com
# @File    : 430_扁平化多级双向链表.go
# @Software: GoLand
-------------------------------------
*/

package main

import (
	"fmt"
)

func main() {
	var root *Node
	root = &Node{Val: 1, Next: &Node{Val: 2, Prev: root}, Child: &Node{Val: 3}}
	res := flatten(root)
	for res != nil {
		fmt.Println(res.Val)
		res = res.Next
	}
}

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}

func flatten(root *Node) *Node {
	if root == nil {
		return root
	}
	head := root
	cur := head
	for cur != nil {
		// 没有子节点
		if cur.Child == nil {
			cur = cur.Next
			continue
		}

		// 含有子节点，找到该子节点的末尾后继节点
		tmp := cur.Child
		for tmp.Next != nil {
			tmp = tmp.Next
		}
		// 连接起来
		tmp.Next = cur.Next
		if cur.Next != nil {
			cur.Next.Prev = tmp
		}

		cur.Next, cur.Child.Prev = cur.Child, cur
		cur.Child = nil // 题目要求
	}
	return head
}
